### PR TITLE
enh(conf): use timeperiod endpoints in UI conf forms (create/update/delete)

### PR DESCRIPTION
**Pull request by a-launois**
## Description

Call endpoints in legacy code for create/update/delete action in UI timeperiod forms

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [x] master

### CODE REVIEW

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 1]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+function insertTimePeriodInAPI(array $ret = []): int|null
+{
+    global $form, $basePath;
+
+    $formData = $ret === [] ? $form->getSubmitValues() : $ret;
+
+    try {
+        $timeperiodId = insertTimeperiodByApi($formData, $basePath);
+
+        return $timeperiodId;
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:408]:
```suggestion
        return insertTimeperiodByApi($formData, $basePath);
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 2]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+}
+
+/**
+ * Make the API request to create a new timeeperiod and return the new ID.
+ *
+ * @param array $formData
+ * @param string $basePath
+ *
+ * @throws LogicException
+ * @throws Exception
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:423]: There is several exceptions that are not in the php docs, you can use this suggestion to fix the alert
```suggestion
 * @throws Throwable
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 3]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+    global $form, $basePath;
+
+    $formData = $ret === [] ? $form->getSubmitValues() : $ret;
+
+    try {
+        $timeperiodId = insertTimeperiodByApi($formData, $basePath);
+
+        return $timeperiodId;
+    } catch (Throwable $th) {
+        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:410]: here you need to write a log with new methods of centreonLog to log the error to improve error handling ;) 
```suggestion
    } catch (Throwable $th) {
        CentreonLog::create()->error(
            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
            message: "Error while inserting timeperiod by api : {$th->getMessage()}",
            customContext: ['form_data' => $formData, 'base_path' => $basePath],
            exception: $th
        );
        
        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 4]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+    global $form, $basePath;
+
+    $formData = $form->getSubmitValues();
+
+    try {
+        updateTimeperiodByApi($formData, $basePath);
+
+        return true;
+    } catch (Throwable $th) {
+        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:488]: here you need to write a log with new methods of centreonLog to log the error to improve error handling ;) 
```suggestion
    } catch (Throwable $th) {
        CentreonLog::create()->error(
            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
            message: "Error while updating timeperiod by api : {$th->getMessage()}",
            customContext: ['form_data' => $formData, 'base_path' => $basePath],
            exception: $th
        );
        
        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 5]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+    }
+}
+
+/**
+ * Make the API request to update a timeeperiod .
+ * @param array $formData
+ * @param string $basePath
+ *
+ * @throws LogicException
+ * @throws Exception
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:500]: There is several exceptions that are not in the php docs, you can use this suggestion to fix the alert
```suggestion
 * @throws Throwable
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 6]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+    $headers = [
+        'Content-Type' => 'application/json',
+        'Cookie' => 'PHPSESSID=' . $_COOKIE['PHPSESSID'],
+    ];
+    $response = $client->request(
+        'PUT',
+        $url,
+        [
+            'headers' => $headers,
+            'body' => json_encode($payload),
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:527]: json_encode can return false on failure, you can use [JSON_THROW_ON_ERROR] (https://www.php.net/manual/en/json.constants.php#constant.json-throw-on-error) flag to improve error handling ;)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 7]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+        ],
+    );
+
+    if ($response->getStatusCode() !== 204) {
+        $content = json_decode($response->getContent(false));
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:532]: json_decode can return null on failure, you can use [JSON_THROW_ON_ERROR] (https://www.php.net/manual/en/json.constants.php#constant.json-throw-on-error) flag to improve error handling ;)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 8]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+function deleteTimePeriodInAPI(array $timeperiods = []): bool
+{
+    global $basePath;
+
+   try {
+        deleteTimeperiodByApi($basePath, $timeperiods);
+
+        return true;
+    } catch (Throwable $th) {
+        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:552]:
```suggestion
    } catch (Throwable $th) {
        CentreonLog::create()->error(
            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
            message: "Error while deleting timeperiod by api : {$th->getMessage()}",
            customContext: ['timeperiods' => $timeperiods, 'base_path' => $basePath],
            exception: $th
        );
        
        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 9]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+
+        return false;
+    }
+}
+
+/**
+ * @param string $basePath
+ * @param int[] $timePeriodIds
+ *
+ * @throws Exception
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:562]: There is several exceptions that are not in the php docs, you can use this suggestion to fix the alert
```suggestion
 * @throws Throwable
```

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 10]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+        $url = $router->generate(
+            'DeleteTimePeriod',
+            $basePath ? ['base_uri' => $basePath, 'id' => $id] : [],
+            UrlGeneratorInterface::ABSOLUTE_URL,
+        );
+
+        $response = $client->request('DELETE', $url, ['headers' => $headers]);
+
+        if ($response->getStatusCode() !== 204) {
+            $content = json_decode($response->getContent(false));
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:586]: json_decode can return null on failure, you can use [JSON_THROW_ON_ERROR] (https://www.php.net/manual/en/json.constants.php#constant.json-throw-on-error) flag to improve error handling ;)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 11]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+            ];
+        }
+        if (in_array($name, array_keys($weekDays), true) && $value !== '') {
+            $days[] = ['day' => $weekDays[$name], 'time_range' => $value];
+        }
+    }
+
+    return [
+        'name' => $formData['tp_name'],
+        'alias' => $formData['tp_alias'],
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:625]: Are you sure that these variables are isset all the time :) ?

**Review by Drago37 — CHANGES_REQUESTED** [review]

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 12]
```diff
@@ -568,3 +386,262 @@ function createTimePeriodsExceptions(array $params): void
...
+        return insertTimeperiodByApi($formData, $basePath);
+    } catch (Throwable $th) {
+        CentreonLog::create()->error(
+            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+            message: "Error while inserting timeperiod by api : {$th->getMessage()}",
+            customContext: ['form_data' => $formData, 'base_path' => $basePath],
+            exception: $th
+        );
+
+        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:415]: Are you sure you want to send the exception  message back to the user ? It is not often user-friendly :)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php (hunk 7 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:542]: Don't forget it please :)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php (hunk 11 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:642]: Ok for me !

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 13]
```diff
@@ -568,3 +386,245 @@ function createTimePeriodsExceptions(array $params): void
...
+        'POST',
+        $url,
+        [
+            'headers' => $headers,
+            'body' => json_encode($payload),
+        ],
+    );
+
+    if ($response->getStatusCode() !== 201) {
+        $content = json_decode($response->getContent(false));
```
[the remark this answers is not in the document]
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php]: ok for me :)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php (hunk 7 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php]: ok for me :)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php (hunk 10 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php]: ok for me :)

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php (hunk 12 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:415]: seen together it's ok

File: centreon/www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 14]
```diff
@@ -568,3 +386,262 @@ function createTimePeriodsExceptions(array $params): void
...
+        return true;
+    } catch (Throwable $th) {
+        CentreonLog::create()->error(
+            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+            message: "Error while updating timeperiod by api : {$th->getMessage()}",
+            customContext: ['form_data' => $formData, 'base_path' => $basePath],
+            exception: $th
+        );
+
+        echo "<div class='msg' align='center'>" . _($th->getMessage()) . '</div>';
```
[the remark this answers is not in the document]
Comment (Drago37) [reply centreon/www/include/configuration/configObject/timeperiod/DB-Func.php:499]: seen together it's ok

**Review by Drago37 — DISMISSED** [review]

**Review by callapa — APPROVED** [review]

**Review by Drago37 — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**